### ISSUE TITLE
Mega: Resolve conflicting Debian packages

### ISSUE THREAD

**Issue by mtlynch** [tiny-pilot/janus-debian#15]
This is a mega ticket to 

## Overview

TinyPilot's Janus Debian package conflicts with three other packages in a way that might create problems for us in the future.

### Janus' package conflicts

Our Janus package currently conflicts with (at least) three other Debian packages:

* libnice10
* libsrtp2-1
* libwebsockets16

libnice10 is the most troublesome because there is a libnice10 package available from standard Raspbian apt repos, whereas I don't think the others are available. Worse, I believe libnice10 is installed by default in Raspbian Bullseye.

The conflict is that in order to function, Janus needs shared libraries from the other packages to be in their expected location. For example, the Janus package places a file at `/usr/lib/arm-linux-gnueabihf/libnice.so.10`. But then the *real* libnice10 package also wants to place a file there. If both packages are trying to place a file in the same location, there's a conflict.

The proper thing to do is to declare the conflict [in the package metadata](https://github.com/tiny-pilot/janus-debian/blob/f350378d03f44a2700295c41ef035501c78d476e/Dockerfile#L170). I didn't understand this before, so I [removed the conflict line](https://github.com/tiny-pilot/janus-debian/pull/7), but it turned out that just pushed the failure later in the install process when apt realizes that another package already owns a file that the Janus package is trying to install.

### Why this is a problem

Package `janus=1.0.1` conflicts with `libnice10` but package `janus=1.0.4` *depends on* libnice10, then we're in pickle because we can't install `janus=1.0.4` without installing `libnice10` first, but we can't install `libnice10` because it would conflict with `janus=1.0.1`.

So we'd have to do something ugly like:

1. Determine if `janus=1.0.1` is installed
2. If yes, remove it
3. Install our custom `libnice10` package
4. Install `janus=1.0.4`

### A. Host our own apt repo

If we host our own solution, I *think* we get more flexibility. Because 

I *think* (but don't know for sure) that if 

#### Advantages

* Should give us more flexibility
* Might be useful if we decide to distribute private packages via a private repo in the future

#### Disadvantage

* More complex than just building one-off .deb files and hosting them in Github releases

### B. Install debian packages one by one

The other possibility is that

**Comment by mtlynch** [comment]
@jdeanwallace @jotaen4tinypilot - Can you review and let me know your thoughts?

I thought this would be small enough for me to handle on my own, but I'm realizing it's larger in scope than I realized, and it's something we need to figure out before we release H.264.

**Comment by jdeanwallace** [comment]
Nice write up. Here are some of my thoughts:

1. From the [Janus repo](https://github.com/meetecho/janus-gateway#dependencies) readme, under dependencies:
    > [libnice](https://libnice.freedesktop.org/) (at least v0.1.16 suggested, v0.1.18 recommended)

    `libnice10` (0.1.16) is available on `buster-backports` and `bullseye` apt repos. Can we consider downgrading libnice10 from 0.1.18 to 0.1.16?

2. > D. Find a trustworthy apt repo for Janus

    When building janus, we currently [set a few flags](https://github.com/tiny-pilot/janus-debian/blob/master/Dockerfile#L89-L93) to remove some unnecessary functionality. I suspect that if we use an already available Janus package, we'll get everything but the kitchen sink. Not a deal breaker though. 

3. In terms of installing janus & friends, option B seems like only a slight improvement on option C. If so, we can always start with the easier option C and then upgrade to B, if needed.

4. > A. Host our own apt repo

    I'm unsure about how smart/dumb `apt` is in finding/resolving dependencies. I'll need to play around with it a bit to see how it behaves.

5. > For users who installed Janus while H264 was in beta, they'll fail the above sequence. They won't be able to install our new libnice10 package because it will conflict with their existing janus package. And they won't be able to install our new janus package because it will have a dependency on our custom build of libnice10, which apt won't know how to find.
    >
    > Maybe we have special logic to remove old versions of janus before installing libnice10, but that's a bit ugly. And then if we ever have to unbundle libsrtp2-1 or libwebsockets16, we're back to the same problem.

    I agree that this would be ugly, but at least these hacks wouldn't live in our code forever because we should be able to push users down a specific download path (i.e., `2.4.1` -> `2.4.8` -> `5.10`) 😬

**Comment by mtlynch** [comment]
Oh, I hadn't really understood what the backports repo was, but from reading about them a little more, they seem like a good option.

If we could install Janus (and possibly other dependencies) from backports, that would be much better than having to maintain our own Debian package, which has been eating up a lot of our dev time.

@jdeanwallace - Can you look into https://github.com/tiny-pilot/ansible-role-ustreamer/issues/77 once you've got the gatekeeper stuff wrapped up?

**Comment by mtlynch** [comment]
Yeah, I'm not sure how it gets installed, but I've definitely had conflicts with libnice10 when I test on Bullseye Lite. Maybe something earlier in the install process installs libnice10 as a side effect?

I think the best option at the moment is to explore whether we can just install using the official packages in the backports repo. Because if that's possible, then we save ourselves a lot of work now and a lot of ongoing maintenance work.
